### PR TITLE
[Notebooks] Transactions for entry creation/editing

### LINKED ISSUE #4744: Unsaved changes to notebook entries can sometimes be lost
#### Summary
Under some circumstances unsaved entries to Notebooks can sometimes be lost. This can occur if a user saves a modification to a notebook (eg. creates a new entry, modifies an existing one, deletes an existing one, etc.) while another user is actively editing their own entry. This will cause the entry that is being actively edited to be lost completely. 

Suspect this is not a regression and has always been the case with multi-user notebooks.

This issue is due to the fact that the notebook model does not get persisted until the user blurs the notebook entry. If a remote change is detected during this time it will just fetch the remote version of the notebook and replace the local notebook model.

#### Impact Check List

- [X] Data loss or misrepresented data?
- [ ] Regression? Did this used to work or has it always been broken?
- [ ] Is there a workaround available?
- [X] Does this impact a critical component?
- [ ] Is this just a visual bug with no functional impact?

### PR DESCRIPTION
**Pull request by jvigliotta**
Closes #4744 

### Describe your changes:

Added functionality to start transactions when created or editing an entry. 

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing document](https://github.com/nasa/openmct/blob/master/CONTRIBUTING.md)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/nasa/openmct/pulls) for the same update/change?
* [x] Is this change backwards compatible? For example, developers won't need to change how they are calling the API or how they've extended core plugins such as Tables or Plots.

### Author Checklist

* [x] Changes address original issue?
* [ ] Unit tests included and/or updated with changes?
* [x] Command line build passes?
* [x] Has this been smoke tested?
* [x] Testing instructions included in associated issue?

### Reviewer Checklist

* [ ] Changes appear to address issue?
* [ ] Changes appear not to be breaking changes?
* [ ] Appropriate unit tests included?
* [ ] Code style and in-line documentation are appropriate?
* [ ] Commit messages meet standards?
* [ ] Has associated issue been labelled unverified? (only applicable if this PR closes the issue)
* [ ] Has associated issue been labelled bug? (only applicable if this PR is for a bug fix)

### CODE REVIEW

File: src/plugins/notebook/components/Notebook.vue [hunk 1]
```diff
@@ -728,6 +734,39 @@ export default {
...
+        },
+        saveTransaction() {
+            const transaction = this.activeTransaction();
+
+            if (!transaction) {
+                return;
+            }
+
+            return transaction.commit()
+                .then(() => {
```
Comment (nikhilmandlik) [inline src/plugins/notebook/components/Notebook.vue:752]: can remove this `.then`?

File: src/plugins/notebook/components/Notebook.vue [hunk 2]
```diff
@@ -728,6 +734,39 @@ export default {
...
+            this.openmct.objects.startTransaction();
+        },
+        saveTransaction() {
+            const transaction = this.activeTransaction();
+
+            if (!transaction) {
+                return;
+            }
+
+            return transaction.commit()
```
Comment (nikhilmandlik) [inline src/plugins/notebook/components/Notebook.vue:751]: missing `;` at end 😄

File: src/plugins/notebook/components/Notebook.vue [hunk 3]
```diff
@@ -728,6 +734,39 @@ export default {
...
+                .then(() => {
+                // committed
+                }).catch(error => {
+                    throw error;
+                }).finally(() => {
+                    this.openmct.objects.endTransaction();
+                });
+        },
+        cancelTransaction() {
+            return new Promise((resolve, reject) => {
```
Comment (nikhilmandlik) [inline src/plugins/notebook/components/Notebook.vue:761]: wondering why do we need to return promise here?

File: src/plugins/notebook/components/Notebook.vue [hunk 4]
```diff
@@ -728,6 +734,39 @@ export default {
             this.selectPage(pageId);
 
             this.syncUrlWithPageAndSection();
+        },
+        activeTransaction() {
+            return this.openmct.objects.getActiveTransaction();
+        },
+        startTransaction() {
```
Comment (nikhilmandlik) [inline src/plugins/notebook/components/Notebook.vue:741]: need to handle (start transaction) if we drop object on 'To start a new entry, click here or drag and drop any object'
`dropOnEntry` function from https://github.com/nasa/openmct/pull/4917/files#diff-65eecdb132430c1ee14c1f6be3551e25a5e64cec0b7370c7d19b47472101d259R391

**Review by nikhilmandlik — CHANGES_REQUESTED** [review]
Looks good, small changes, and need to handle one missing case

File: src/plugins/notebook/components/Notebook.vue (hunk 1 above)
Comment (jvigliotta) [reply src/plugins/notebook/components/Notebook.vue:752]: But, but, what about that comment???? Yes. I will remove.

File: src/plugins/notebook/components/Notebook.vue (hunk 2 above)
Comment (jvigliotta) [reply src/plugins/notebook/components/Notebook.vue:751]: we've got the catch and finally after it, it's at the end of those

File: src/plugins/notebook/components/Notebook.vue (hunk 3 above)
Comment (jvigliotta) [reply src/plugins/notebook/components/Notebook.vue:761]: we definitely don't need to, I'll remove, it was a holdover from pulling the code from transactions.

File: src/plugins/notebook/components/Notebook.vue (hunk 4 above)
Comment (jvigliotta) [reply src/plugins/notebook/components/Notebook.vue:741]: We actually don't need to do this, it's started in NotebookEntry.vue when an item is dropped in Notebook.vue

**Review by akhenry — APPROVED** [review]

**Review by akhenry — CHANGES_REQUESTED** [review]
I think this looks good, we do just need to be careful not to commit or cancel transactions that we didn't start though.

**Comment by akhenry** [comment]
Sorry, accidentally approved when I meant to request changes :)

File: src/plugins/notebook/components/Notebook.vue [hunk 5]
```diff
@@ -746,6 +752,35 @@ export default {
...
+            this.openmct.objects.startTransaction();
```
[the remark this answers is not in the document]
Comment (jvigliotta) [reply src/plugins/notebook/components/Notebook.vue]: Gotcha! Done.

File: src/plugins/notebook/components/Notebook.vue [hunk 6]
```diff
@@ -746,6 +752,35 @@ export default {
...
+        },
+        saveTransaction() {
+            const transaction = this.activeTransaction();
+
+            if (!transaction) {
+                return;
+            }
+
+            return transaction.commit()
```
[the remark this answers is not in the document]
Comment (jvigliotta) [reply src/plugins/notebook/components/Notebook.vue:771]: Also, Gotcha! Done.

File: src/plugins/notebook/components/Notebook.vue [hunk 7]
```diff
@@ -746,6 +752,35 @@ export default {
...
+                .catch(error => {
+                    throw error;
+                }).finally(() => {
+                    this.openmct.objects.endTransaction();
+                });
+        },
+        cancelTransaction() {
+            const transaction = this.activeTransaction();
+            transaction.cancel()
```
[the remark this answers is not in the document]
Comment (jvigliotta) [reply src/plugins/notebook/components/Notebook.vue]: Annnnd, Gotcha! Done.

**Review by akhenry — DISMISSED** [review]
### Reviewer Checklist

1. Changes appear to address issue? Y
2. Appropriate unit tests included? N/A. Small bug fix
3. Code style and in-line documentation are appropriate? Y
4. Commit messages meet standards? Y
5. Has associated issue been labelled `unverified`? (only applicable if this PR closes the issue)

**Review by nikhilmandlik — APPROVED** [review]
LGTM

**Review by akhenry — APPROVED** [review]

**Pull request merged**